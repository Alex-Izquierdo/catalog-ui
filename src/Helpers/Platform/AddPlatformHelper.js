import React from 'react';
import { getUserApi } from '../Shared/userLogin';

const api = getUserApi();

export function addPlatform(providerData) {
  return api.addProvider(providerData).then((data) => {
    console.log('Add provider API called successfully. Returned data: ' + data);
    return data;
  }, (error) => {
    console.error(error);
  });
}
