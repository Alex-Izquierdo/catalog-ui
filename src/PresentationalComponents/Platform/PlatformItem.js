import React from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import './platformitem.scss';
import propTypes from 'prop-types';
import CatItemSvg from '../../assets/images/vendor-openshift.svg';
import ImageWithDefault from '../Shared/ImageWithDefault';
import { PlatformStore, fetchPlatformItems } from '../../Store/Actions/PlatformActions';
import { hideModal, showModal } from '../../Store/Actions/MainModalActions';
import { GridItem, Card, CardHeader, CardBody, CardFooter, Checkbox } from '@patternfly/react-core';
import { Dropdown, DropdownItem, DropdownPosition, DropdownToggle } from '@patternfly/react-core';
import itemDetails from '../../PresentationalComponents/Shared/CardCommon';
import { bindMethods, consoleLog } from '../../Helpers/Shared/Helper';
import CardCheckbox from '../Shared/CardCheckbox';

const TO_DISPLAY = [ 'description' ];

const mapDispatchToProps = dispatch => {
    return {
        hideModal: () => dispatch(hideModal()),
        showModal: (modalProps, modalType) => {
            dispatch(showModal({ modalProps, modalType }));
        }
    };
};

class PlatformItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = { isOpen: true, showMenu: false, isEditMode: false };
    };

    onSelect = (event) => {
        consoleLog('This is the selected state:', this.state);
        consoleLog('This is the selected event:', event);

        this.props.showModal({
            open: true,
            itemdata: this.props,
            closeModal: this.props.hideModal
        }, 'addportfolio');

        this.setState({
            ...this.state,
            isOpen: !this.state.isOpen
        });
    };

    isChecked = () => {
        if (this.props.isChecked)
        {return this.props.isChecked(this.props.id);}
        else
        {return false;}
    }

    render() {
        return (
            <GridItem key={ this.props.id } GridItem sm={ 6 } md={ 4 } lg={ 4 } xl={ 3 }>
                <Card key={ this.props.id }>
                    <CardHeader className="card_header">
                        <ImageWithDefault src={ this.props.imageUrl || CatItemSvg } defaultSrc={ CatItemSvg } width="50" height="50" />
                        { this.props.isEditMode && <CardCheckbox id={ this.props.id } checked={ this.isChecked } onChange={ this.props.onCheckboxClick } /> }
                    </CardHeader>
                    <CardBody className="card_body">
                        <h4>{ this.props.name }</h4>
                        { itemDetails(this.props, TO_DISPLAY) }
                    </CardBody>
                    <CardFooter/>
                </Card>
            </GridItem>
        );
    };
}

PlatformItem.propTypes = {
    history: propTypes.object,
    showModal: propTypes.func,
    hideModal: propTypes.func,
    imageUrl: propTypes.string,
    id: propTypes.string,
    name: propTypes.string,
    isEditMode: propTypes.string,
    checkedItems: propTypes.object,
    isChecked: propTypes.string,
    onCheckboxClick: propTypes.func,
    onCheckboxClick: propTypes.func
};

export default withRouter(PlatformItem);

