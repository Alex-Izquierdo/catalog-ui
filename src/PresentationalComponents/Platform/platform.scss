@import "../../App";
@import "../Shared/toolbarschema";

.card_style_with_hover {
  background-color: white;
}

.card_style_with_hover:hover {
  background-color:darkgrey;
  opacity: 0.8
}
