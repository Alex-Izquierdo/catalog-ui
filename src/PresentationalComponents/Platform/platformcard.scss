@import "../../App.scss";

.pcard_header {
  height: 60px;
}

.pcard_body {
  height: 140px;
  overflow: hidden;
  font-size: small;
}

.pcard_element {
  text-align: left;
  font-size: small;
}
