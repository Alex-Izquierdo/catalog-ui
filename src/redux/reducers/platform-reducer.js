import {
  FETCH_PLATFORM,
  FETCH_PLATFORMS,
  FETCH_PLATFORM_ITEMS,
  FETCH_PLATFORM_ITEM,
  FILTER_PLATFORM_ITEMS,
  FETCH_MULTIPLE_PLATFORM_ITEMS,
  SET_LOADING_STATE,
  FETCH_PLATFORM_INVENTORIES,
  SET_INVENTORIES_LOADING_STATE
} from '../action-types';

// Initial State
export const platformInitialState = {
  isPlatformDataLoading: false,
  selectedPlatform: {},
  platforms: [],
  platformItems: {},
  platformInventories: [],
  platformItem: {},
  platform: {},
  filterValue: ''
};

// rename isPlatformLoading.. to isLoading so we can use common action for loading states

const setLoadingState = (state, { payload = true }) => ({
  ...state,
  isPlatformDataLoading: payload
});
const setPlatforms = (state, { payload }) => ({
  ...state,
  platforms: payload,
  isPlatformDataLoading: false
});
const setPlatformItems = (state, { payload, meta: { platformId } }) => ({
  ...state,
  platformItems: { ...state.platformItems, [platformId]: payload },
  isPlatformDataLoading: false
});
const setMultiplePlatformItems = (state, { payload }) => ({
  ...state,
  platformItems: { ...state.platformItems, ...payload },
  isPlatformDataLoading: false
});
const setPortfolioItems = (state, { payload }) => ({
  ...state,
  portfolioItem: payload,
  isPlatformDataLoading: false
});
const selectPlatform = (state, { payload }) => ({
  ...state,
  selectedPlatform: payload,
  isLoading: false
});
const filterPlatformItems = (state, { payload }) => ({
  ...state,
  filterValue: payload
});
const setInventoriesDataLoadingState = (state, { payload = true }) => ({
  ...state,
  isInventoriesDataLoading: payload
});
const setPlatformInventories = (state, { payload }) => ({
  ...state,
  platformInventories: { ...state.platformInventories, ...payload },
  isInventoriesDataLoading: false
});

export default {
  [`${FETCH_PLATFORMS}_PENDING`]: setLoadingState,
  [`${FETCH_PLATFORMS}_FULFILLED`]: setPlatforms,
  [`${FETCH_PLATFORM_ITEMS}_PENDING`]: setLoadingState,
  [`${FETCH_PLATFORM_ITEMS}_FULFILLED`]: setPlatformItems,
  [`${FETCH_PLATFORM_ITEM}_PENDING`]: setLoadingState,
  [`${FETCH_PLATFORM_ITEM}_FULFILLED`]: setPortfolioItems,
  [`${FETCH_PLATFORM}_PENDING`]: setLoadingState,
  [`${FETCH_PLATFORM}_FULFILLED`]: selectPlatform,
  [`${FILTER_PLATFORM_ITEMS}_FULFILLED`]: filterPlatformItems,
  [`${FETCH_MULTIPLE_PLATFORM_ITEMS}_FULFILLED`]: setMultiplePlatformItems,
  [SET_LOADING_STATE]: setLoadingState,
  [`${FETCH_PLATFORM_INVENTORIES}_PENDING`]: setInventoriesDataLoadingState,
  [`${FETCH_PLATFORM_INVENTORIES}_FULFILLED`]: setPlatformInventories,
  [SET_INVENTORIES_LOADING_STATE]: setInventoriesDataLoadingState
};
