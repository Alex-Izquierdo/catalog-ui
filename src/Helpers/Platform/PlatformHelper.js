import React from 'react';
import PlatformItem from '../../PresentationalComponents/Platform/PlatformItem';
import { getTopologicalUserApi } from '../Shared/userLogin';

const api = getTopologicalUserApi();

export function getPlatforms() {
  return api.listSources().then((data) => {
    console.log('listSources API called successfully. Returned platforms: ' + data);
    return data;
  }, (error) => {
    console.error(error);
  });
}

export function getPlatformItems(apiProps) {
  console.log('getPlatformItems called with : ' + apiProps.platform);
  let apiPromise = null;
  if( apiProps && apiProps.platform) {
    // TODO - replace with offerings per source when available
    apiPromise = api.listServiceOfferings();
  }
  else {
    apiPromise = api.listServiceOfferings();
  }

  return apiPromise.then((data) => {
      console.log('API called successfully. Returned data: ' + data);
      return processPlatformItems(data);
  }, (error) => {
      console.error(error);
  });
}

function processPlatformItems(items) {
  let platformItems = [];
  items.forEach( function(item, row, _array) {
    let newRow = processPlatformItem(row, item);
    platformItems.push(newRow);
  });

  return {platformItems};
}

function processPlatformItem(key, data) {
    return <PlatformItem key={key} {...data} />;
}

