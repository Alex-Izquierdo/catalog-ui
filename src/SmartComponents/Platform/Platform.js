import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Title } from '@patternfly/react-core';
import { Main } from '@red-hat-insights/insights-frontend-components';
import { fetchSelectedPlatform, fetchPlatformItems } from '../../redux/Actions/PlatformActions';
import ContentGallery from '../../SmartComponents/ContentGallery/ContentGallery';
import PlatformToolbar from '../../PresentationalComponents/Platform/PlatformToolbar';
import PlatformItem from '../../PresentationalComponents/Platform/PlatformItem';
import { scrollToTop } from '../../Helpers/Shared/helpers';
import './platform.scss';

class Platform extends Component {
  state = {
    platformId: '',
    filteredItems: []
  };

  fetchData(apiProps) {
    this.props.fetchSelectedPlatform(apiProps);
    this.props.fetchPlatformItems(apiProps);
  }

  componentDidMount() {
    this.fetchData(this.props.match.params.id);
    scrollToTop();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      this.fetchData(this.props.match.params.id);
      scrollToTop();
    }
  }

  render() {
    let filteredItems = {
      items: this.props.platformItems.map(data => <PlatformItem key={ data.id } { ...data } />),
      isLoading: this.props.isPlatformDataLoading
    };

    let title = this.props.platform ? this.props.platform.name : '';

    return (
      <Main style={ { marginLeft: 0, paddingLeft: 0, paddingTop: 0 } }>
        <PlatformToolbar/>
        <div style={ { marginLeft: 25, paddingTop: 40 } }>
          { title &&  (<Title size={ '2xl' } > { title }</Title>) }
        </div>
        <ContentGallery { ...filteredItems }/>
      </Main>
    );
  }
}

const mapStateToProps = ({ platformReducer: { selectedPlatform, platformItems, isPlatformDataLoading }}) => ({
  platform: selectedPlatform,
  platformItems: selectedPlatform && platformItems[selectedPlatform.id],
  isPlatformDataLoading: !selectedPlatform || isPlatformDataLoading
});

const mapDispatchToProps = dispatch => ({
  fetchSelectedPlatform: platformId => dispatch(fetchSelectedPlatform(platformId)),
  fetchPlatformItems: apiProps => dispatch(fetchPlatformItems(apiProps))
});

Platform.propTypes = {
  filteredItems: PropTypes.object,
  isPlatformDataLoading: PropTypes.bool,
  match: PropTypes.object,
  fetchPlatformItems: PropTypes.func.isRequired,
  fetchSelectedPlatform: PropTypes.func,
  platform: PropTypes.shape({
    name: PropTypes.string
  }),
  platformItems: PropTypes.array
};

Platform.defaultProps = {
  platformItems: []
};

export default connect(mapStateToProps, mapDispatchToProps)(Platform);
