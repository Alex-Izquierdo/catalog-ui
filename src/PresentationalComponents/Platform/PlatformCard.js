import React from 'react';
import propTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { GridItem, Card, CardHeader, CardBody, CardFooter } from '@patternfly/react-core';
import DefaultPlatformImg from '../../assets/images/platform-default.svg';
import OpenshiftPlatformImg from '../../assets/images/platform-openshift.svg';
import AmazonPlatformImg from '../../assets/images/platform-amazon.png';
import ImageWithDefault from '../Shared/ImageWithDefault';
import ItemDetails from '../Shared/CardCommon';
import './platformcard.scss';

const TO_DISPLAY = [ 'description', 'modified' ];

// TO DO - use webpack to load all images
const platformTypeImg = (sourceTypeId) => {
  switch (sourceTypeId) {
    case '1':
      return OpenshiftPlatformImg;
    case '2':
      return AmazonPlatformImg;
    default:
      return DefaultPlatformImg;
  }
};

const PlatformCard = ({ name, id, ...props }) => (
  <GridItem sm={ 6 } md={ 4 } lg={ 4 } xl={ 3 }>
    <Link to={ `/platform/${id}` } className="card-link">
      <Card key={ id }>
        <CardHeader className="pcard_header">
          <ImageWithDefault src={ platformTypeImg(props.source_type_id || '0') } defaultSrc={ DefaultPlatformImg } width="80" height="40"/>
        </CardHeader>
        <CardBody className="pcard_body">
          <h4>{ name }</h4>
          <ItemDetails { ...{ name, ...props } } toDisplay={ TO_DISPLAY } />
        </CardBody>
        <CardFooter/>
      </Card>
    </Link>
  </GridItem>
);

PlatformCard.propTypes = {
  history: propTypes.object,
  imageUrl: propTypes.string,
  name: propTypes.string,
  source_type_id: propTypes.string,
  id: propTypes.string
};

export default PlatformCard;
