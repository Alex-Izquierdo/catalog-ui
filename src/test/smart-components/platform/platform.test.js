import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Platform from '../../../smart-components/platform/platform';
import { platformInitialState } from '../../../redux/reducers/platform-reducer';
import { Provider } from 'react-redux';
import { Route, MemoryRouter } from 'react-router-dom';

describe('<Platform />', () => {
  let initialProps;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;
  let initialState;

  const ComponentWrapper = ({ store, initialEntries = ['/foo'], children }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>
        <Route path="/platforms">{children}</Route>
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {};
    mockStore = configureStore(middlewares);
    initialState = {
      platformReducer: {
        ...platformInitialState,
        selectedPlatform: {
          id: '1',
          name: 'Foo'
        },
        platforms: [
          {
            id: '1',
            name: 'Foo'
          }
        ]
      }
    };
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(
      <ComponentWrapper store={store} initialEntries={['/platforms']}>
        <Platform store={mockStore(initialState)} {...initialProps} />
        );
      </ComponentWrapper>
    ).find(Platform);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
