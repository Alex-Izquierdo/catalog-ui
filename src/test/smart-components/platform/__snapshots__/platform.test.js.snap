// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Platform /> should render correctly 1`] = `
<Platform
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
